lib/pos_mempool.go

```diff
@@ -0,0 +1,490 @@
+	// The ledger keeps track of how much each user would have spent in fees across all their transactions in the mempool.
+	ledger *BalanceLedger
+	// persister is responsible for interfacing with the database. The persister backs up mempool transactions so not to
+	// lose them when node reboots. The persister also retrieves transactions from the database when the node starts up.
+	// The persister runs on its dedicated thread and events are used to notify the persister thread whenever
+	// transactions are added/removed from the mempool. The persister thread then updates the database accordingly.
+	persister *MempoolPersister
+	// nonceTracker is responsible for keeping track of a (public key, nonce) -> Txn index. The index is useful in
+	// facilitating a "replace by higher fee" feature. This feature gives users the ability to replace their existing
+	// mempool transaction with a new transaction having the same nonce but higher fee.
+	nonceTracker *NonceTracker
+
+	// latestBlockView is used to check if a transaction is valid before being added to the mempool. The latestBlockView
+	// checks if the transaction has a valid signature and if the transaction's sender has enough funds to cover the fee.
+	// The latestBlockView should be updated whenever a new block is added to the blockchain via UpdateLatestBlock.
+	// PosMempool only needs read-access to the block view. It isn't necessary to copy the block view before passing it
+	// to the mempool.
+	latestBlockView *UtxoView
+	// latestBlockNode is used to infer the latest block height. The latestBlockNode should be updated whenever a new
+	// block is added to the blockchain via UpdateLatestBlock.
+	latestBlockHeight uint64
+}
+
+// PosMempoolIterator is a wrapper around FeeTimeIterator, modified to return MsgDeSoTxn instead of MempoolTx.
+type PosMempoolIterator struct {
+	it *FeeTimeIterator
+}
+
+func (it *PosMempoolIterator) Next() bool {
+	return it.it.Next()
+}
+
+func (it *PosMempoolIterator) Value() (*MsgDeSoTxn, bool) {
+	txn, ok := it.it.Value()
+	if txn == nil || txn.Tx == nil {
+		return nil, ok
+	}
+	return txn.Tx, ok
+}
+
+func (it *PosMempoolIterator) Initialized() bool {
+	return it.it.Initialized()
+}
+
+func NewPosMempoolIterator(it *FeeTimeIterator) *PosMempoolIterator {
+	return &PosMempoolIterator{it: it}
+}
+
+func NewPosMempool(params *DeSoParams, globalParams *GlobalParamsEntry, latestBlockView *UtxoView,
+	latestBlockHeight uint64, dir string) *PosMempool {
+	return &PosMempool{
+		status:            PosMempoolStatusNotRunning,
+		params:            params,
+		globalParams:      globalParams,
+		dir:               dir,
+		latestBlockView:   latestBlockView,
+		latestBlockHeight: latestBlockHeight,
+	}
+}
+
+func (dmp *PosMempool) Start() error {
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	if dmp.IsRunning() {
+		return nil
+	}
+
+	// Setup the database.
+	mempoolDirectory := filepath.Join(dmp.dir, "mempool")
+	opts := DefaultBadgerOptions(mempoolDirectory)
```

Comment:
I'm wondering if we want to deal with the memory requirements of opening a new badger. But I think if we can keep it on the DEFAULT options it's pretty low right? How much memory does a default badger instance use if you're regularly writing a gig to it?